draft-ietf-ppm-dap.md

```diff
@@ -1338,11 +1362,11 @@ following generic checks.
       store the set of IDs of processed reports. However, implementations may
       find it helpful to track additional information, like the timestamp, so
       that the storage used for anti-replay can be sharded efficiently.
+    * This also covers report re-use for the Prio3 VDAF because its aggregation
+      parameter is nil.
 
-1. Check if the report has never been aggregated but is contained by a batch
-   that has been collected. If this check fails, the input share MUST be marked
-   as invalid with the error `batch_collected`. This prevents additional reports
-   from being aggregated after its batch has already been collected.
+1. VDAFs may require specific additional checks. For example, Poplar1 requires
+   that reports are aggregated at most once at each level of the IDPF tree.
```

Comment:
I don't think a prescription like this suffices. This should be written in terms of some function exposed by VDAFs, so that DAP implementations can remain generic. So I believe we're blocked on a corresponding VDAF change here.

Reply:
Agreed, this is supposed to be a bandaid. Until we have the VDAF change we have the only requirement that we know of covered here.

Reply:
I actually think that @simon-friedberger is more or less on the right track here, however I agree we want to be generic. Here is where we have converged on the VDAF spec so far (https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/161, please feel free to offer feeedback there as well):

> A VDAF *MAY* specify the "validity scope" of each aggregation parameter. Protocols using a VDAF (like DAP) *MUST* ensure that for each validity scope, a report is aggregated at most once with an aggregation parameter of that validity scope. If no validity scope is specified, a report *MUST NOT* be aggregated more than once with a given aggregation parameter.

This provides what we need for the current VDAFs:

* For Poplar1 the validity scope is the IDPF tree level, which is included in the aggregation parameter; the language here seems sufficient to ensure we don't evaluate at the same level more than once.

* For Prio3 we have no validity scope, however the only valid aggregation parameter is the empty string. The language implies we don't aggregate the same report more than once [1].

So how does this translate into normative text for input share validation? Maybe something like this:

> Check if the input share was previously aggregated with the aggregation parameter's validity scope {{reference-to-definition-in-the-vdaf-draft}}. If so, the Aggregator *MUST* mark the input share as invalid with the error "XXX".

cc/ @schoppmp

[1] The attack we're considering here: We want to make sure the attacker can't query the same input/proof on multiple distinct query randomnesses. If we can ensure by some other means that both the nonce and verify_key can't change overtime, then this may not be strictly necessary. Since we haven't really settled the details here, making sure we aggregate each report at most once seems like a reasonable conservative choice.

Reply:
I agree with Tim, since this is VDAF specific it should be implemented on the VDAF side.
How about:
> Check if the input share was not previously aggregated with the aggregation parameter's validity scope by calling `VDAF.is_valid(report_id, agg_param)`. Otherwise, the Aggregator MUST mark the input share as invalid with the error "invalid_aggregation_parameter".

Reply:
Something like that, yeah. However, `VDAF.is_valid()` would also need to take as input the the agg params that have been used for report_id in the past.

Reply:
@schoppmp Yes, that sounds good, so `validity_scope = VDAF.validity_scope(agg_param)` would extract parts of the agg_param (or otherwise construct) an opaque bitstring and DAP ensures that (report_id, validity_scope) has never been aggregated before, right?

Reply:
Yup, that sounds right to me.

Reply:
IIUC we want to write this section in terms of the interface being added in https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/161